### PR TITLE
Don't store typechain output in git

### LINKED ISSUE #47: Add contract typing to gitignore
> Is there a reason not to add `/src/contract-types` to the `.gitignore` on this project? Entire directory is generated (ie, non-source) files. They clog up the diff count!

I think that's a valid point ([and the correct approach to follow according to typechain](https://github.com/dethcrypto/TypeChain#q-should-i-commit-generated-files-to-my-repository)). I'll address this in a separate PR!

_Originally posted by @lalexgap in https://github.com/statechannels/SAFE-protocol/issues/45#issuecomment-984022839_

### PR DESCRIPTION
**Pull request by lalexgap**
Fixes #47 

Having typechain output checked in means we're constantly dealing with generated files. This PR:
- moves the typechain output out of `src`
- Adds a `.gitignore` entry for typechain output
- Add a postinstall hook to run typechain

### CODE REVIEW

File: test/l2.test.ts [hunk 1]
```diff
@@ -21,7 +21,7 @@ let l2Contract: L2Contract;
 describe("L2 Contract", function () {
   beforeEach(async () => {
     const l2Deployer = await ethers.getContractFactory("L2Contract", bobWallet); // Bob deploys
-    l2Contract = (await l2Deployer.deploy()).connect(aliceWallet); // Alice sends all further txs. She'll pay the gas in this example, and be msg.sender when it counts.
+    l2Contract = (await l2Deployer.deploy()).connect(aliceWallet) as L2Contract; // Alice sends all further txs. She'll pay the gas in this example, and be msg.sender when it counts.
```
Comment (lalexgap) [inline test/l2.test.ts:24]: Not sure why the cast is necessary here but I didn't feel like spending the time to figure it out.

Comment (NiloCK) [reply test/l2.test.ts:24]: I think the contracts need to be instantiated from the typechain factory methods. [Example](https://github.com/dethcrypto/TypeChain/blob/156c7444140ef6092430eab8a87ffbca72277237/examples/ethers-v5/src/index.ts#L2).

But really, the current version is about the same level of boilerplate for the same (or very similar) effect. The risk is that we make a false `as` assertion.

**Review by NiloCK — APPROVED** [review]
Great.

**Comment by lalexgap** [comment]
>I think the contracts need to be instantiated from the typechain factory methods. Example.

Ahh nice! I find it slightly strange that `typechain` overrides `ethers` `getContractFactory` but doesn't seem to strictly type it.

File: test/l2.test.ts (hunk 1 above)
Comment (lalexgap) [reply test/l2.test.ts]: Updated in 46cada5

**Pull request merged**